File: nvpl_blas/c/ztpmv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_ztpmv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t incX = 1;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasUpper;
    enum CBLAS_TRANSPOSE trans = CblasNoTrans;
    enum CBLAS_DIAG diag = CblasNonUnit;
    nvpl_dcomplex_t * AP = NULL;
    nvpl_dcomplex_t * X = NULL;
    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_ap = (N * (N + 1)) / 2;

    printf("\nExample: cblas_ztpmv for a matrix-vector product using a triangular packed matrix\n\n");
    printf("#### args: n=%" PRId64 ", incX=%" PRId64 ", order=%c, uplo=%c, trans=%c, diag=%c\n",
           (int64_t)N, (int64_t)incX, order_to_char(order), uplo_to_char(uplo), transpose_to_char(trans),
           diag_to_char(diag));

    // allocate memory
    AP = (nvpl_dcomplex_t *)malloc(len_ap * sizeof(nvpl_dcomplex_t));
    X = (nvpl_dcomplex_t *)malloc(len_x * sizeof(nvpl_dcomplex_t));

    // fill data
    fill_zvector(AP, len_ap, 1);
    fill_zvector(X, N, incX);

    // print input data
    print_zvector(AP, len_ap, 1, "AP");
    print_zvector(X, N, incX, "X");

    // call cblas_ztpmv
    cblas_ztpmv(order, uplo, trans, diag, N, AP, X, incX);

    // print result
    printf("\nAfter ztpmv operation:\n");
    print_zvector(X, N, incX, "X");

    // release memory
    free(AP);
    free(X);
    return 0;
}

File: nvpl_blas/f77/ztpmv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     ZTPMV 
!
!******************************************************************************
      program   ZTPMV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      integer         n, incx, i
      parameter       (n=2, incx=1)
      integer         xsize, apsize
      parameter       (xsize=1+(n-1)*abs(incx),
     &                 apsize=n*(n+1)/2)
      double complex  x(xsize), ap(apsize)

*       External Subroutines
      external        print_zvector, fill_zvector
      character*1     uplo, trans, diag
      parameter       (uplo='U', trans='N', diag='N')
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, uplo, trans, diag, incx
      call fill_zvector(x,n,incx)
      call fill_zvector(ap,apsize,1)
      call print_zvector(x,n,incx,'X')
      call print_zvector(ap,apsize,1,'AP')
*
*      Call ZTPMV subroutine
      call ZTPMV(uplo,trans,diag,n,ap,x,incx)
*
      print*
      print 102
      call print_zvector(x,n,incx,'X')

      stop
 99   format('Example: ZTPMV for a matrix-vector product using a'
     &       ' triangular packed matrix')
 100  format('#### args: n=',i1,'uplo=',a1,', trans=',a1,
     &       ', diag=',a1,', incx=',i1)
 102  format('The result of ZTPMV: ')
      end
